janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java

```diff
@@ -236,16 +533,168 @@ public boolean evaluateRaw(String value, String term) {
+            return Text.NOT_CONTAINS_FUZZY;
+        }
+
         @Override
         public String toString() {
             return "textContainsFuzzy";
         }
 
+    },
+
+    /**
+     * Whether the text doesnt contain a token is at X Levenshtein of a token (case insensitive)
+     * with X=:
+     * - 0 for strings of one or two characters
+     * - 1 for strings of three, four or five characters
+     * - 2 for strings of more than five characters
+     */
+    NOT_CONTAINS_FUZZY {
+        @Override
+        public boolean test(Object value, Object condition) {
+            this.preevaluate(value, condition);
+            return value != null && evaluateRaw(value.toString(), (String) condition);
+        }
+
+        @Override
+        public boolean evaluateRaw(String value, String term) {
+            for (String token : tokenize(value.toLowerCase())) {
+                if (isFuzzy(term.toLowerCase(), token)) return false;
+            }
+            return true;
+        }
+
+        @Override
+        public boolean isValidCondition(Object condition) {
+            return condition instanceof String && StringUtils.isNotBlank(condition.toString());
+        }
+
+        @Override
+        public boolean hasNegation() {
+            return true;
+        }
+
+        @Override
+        public JanusGraphPredicate negate() {
+            return Text.CONTAINS_FUZZY;
+        }
+
+        @Override
+        public String toString() {
+            return "textNotContainsFuzzy";
+        }
+
+    },
+
+    /**
+     * Whether the text contains a given token sequence in the text (case insensitive)
+     */
+    CONTAINS_PHRASE {
+
+        @Override
+        public boolean test(Object value, Object condition) {
+            this.preevaluate(value, condition);
+            return value != null && evaluateRaw(value.toString(), (String) condition);
+        }
+
+        @Override
+        public boolean evaluateRaw(String value, String terms) {
+            List<String> valueTerms = tokenize(value.trim().toLowerCase());
+            List<String> tokenTerms = tokenize(terms.trim().toLowerCase());
+            if (!terms.isEmpty() && tokenTerms.isEmpty()) return false;
+            return (Collections.indexOfSubList(valueTerms, tokenTerms) != -1);
+        }
+
+        @Override
+        public boolean isValidCondition(Object condition) {
+            return condition instanceof String && StringUtils.isNotBlank((String) condition);
+        }
+
+        @Override
+        public boolean hasNegation() {
+            return true;
+        }
+
+        @Override
+        public JanusGraphPredicate negate() {
+            return Text.NOT_CONTAINS_PHRASE;
+        }
+
+        @Override
+        public String toString() {
+            return "textContainsPhrase";
+        }
+    },
+
+    /**
+     * Whether the text does not contains a given token sequence in the text (case insensitive)
+     */
+    NOT_CONTAINS_PHRASE {
```

Comment:
Can you add the new predicates to search-predicates.md?

Reply:
Will do.

I did have a question on text-search.md, as Ive only implemented the Elastic index backend. Should I update that page as well, or leave it out because the support across the other indexing backends is limited?

Reply:
I think ideally, we should have it implemented across all indexing backends if possible. If you cannot do it now, maybe you could mention in text-search.md that those predicates only work for ES backend.

Reply:
Updated the text-search.md file accordingly to describe the added functionality to Elastic for both text and string predicates.

As for the CLA, Ive kicked off the process to get approval/whitelist from our corporate CLA manager. Hopefully that is resolved soon.